CI: Automatically assign a committer to PRs not opened by committers

LilithHafner opened the pull request:
This is the first step in implementing @StefanKarpinski's [state machine](https://discourse.julialang.org/t/suggestion-to-slightly-improve-julia-development/50916/82), as prototyped here: https://github.com/LilithHafner/AutomationTesting/.

The goal is to make sure all PRs have a committer tracking them, but without adding much cognitive load to committers. This alone will be IMO slightly helpful but not super impactful. Some next steps that will increase impact are
- tracking when responsibility for continued progress lies with the PR author vs with the committer
- gentle reminders to the person responsible, with an option to de-assign self and/or close the PR
- automatically close PR after a long period of inactivity while the PR author is responsible for progress
- automatically re-assign a new committer after a long period of inactivity while the committer is responsible

I chose to use random assignment to members of the committers team because 
- The github suggested reviewer option is not available via API
- Using git blame (or "suggested reviewer") will have a tendency to assign the most busy people to the most PRs which is not good

One technical aspect I still need feedback or suggestions about is how to access the committers list. That list is currently only available to members of the JuliaLang organization and `GITHUB_TOKEN` does not, by default have `org:read` permissions. Some options are
1. give `GITHUB_TOKEN` `org:read`. This could plausibly leak any information accessible via `org:read` (notably JuliaLang membership)
2. use a personal access token (similar to 1)
3. make the `committers` team public (which AFAICT still does not allow API access without organization scope) and access that list through hard to maintain web scraping
4. hardcode and maintain a list of committers.

LilithHafner committed:
Create AssignReviewer.yml

LilithHafner committed:
Whitespace

KristofferC commented:
Seems like this will just add noise to the PRs people are actually assigned to vs the ones they got randomly assigned to? Perhaps it should be opt-in to be one of the people who get PRs randomly assigned to them?

oscardssmith commented:
I think we want assigned reviewer to be opt in rather than all committers. We want to give committer permission widely, and IMO that shouldn't come with the responsibility of tracking PRs other than your own.

LilithHafner commented:
Okay, I'll implement a separately maintained opt-in list.

StefanKarpinski commented:
I don't thing that committers list should be private, so I have no problem with giving the GitHub API token access to it. In general, I would much prefer if open source org membership were public at least by default, but unfortunately org membership is private by default with no way to change that default (afaik), and worse there's no way for an org admin to change anyone else's public/private status, so you literally have to pester each person to change their status in each org to be public. It's very annoying and leads to orgs that look completely empty and have no way to figure out who to contact about getting things done. I get this design for company orgs but it's the opposite of how it should work for open source orgs, imo. Anyway, long rant is to say: I think that commiters should be public information anyway and certainly don't mind the GitHub api token having access to it. We could also have a file somewhere that is a list of committers to choose from when assigning people to PRs.

A way to think about this random assignment: you don't have to remain responsible for the PR—after all there's a decent chance that you don't have the requisite knowledge if assignment is random—but you do have to decide who else might be willing and able to take responsibility for it. Yes, it's a bit of noise, but not much. I'd be happy to get pinged on a random subset of PRs and either just merge them if I can, eg it's a simple improvement and just needs someone to look at it, or I happen to be a person with the necessary expertise, or find someone else who I think can take care of the PR.

LilithHafner commented:
Tagging triage to talk about this specific PR and also the larger plan

DilumAluthge commented:
Regarding the specific implementation in this PR:

For this specific task (automatically assign a member of some team `@JuliaLang/team_X`), I think we don't need a GitHub Action. Instead, I think we can implement the same logic using built-in GitHub functionality.

I agree that someone should be able to be a member of the `@JuliaLang/committers` team without also being part of the "auto-assign" pool. So I agree that we should use a different GitHub team.

Here is what I would propose:

1. Create a new GitHub team.
    - I have created a new GitHub team named `@JuliaLang/committers-available-to-review` (https://github.com/orgs/JuliaLang/teams/committers-available-to-review).
    - As a starting point, I have added @LilithHafner, @StefanKarpinski, and myself as members of `@JuliaLang/committers-available-to-review`.
    - People can
2. Set the `@JuliaLang/committers-available-to-review` team as the default CODEOWNER for the entire `JuliaLang/julia` repo.
    - [ ] https://github.com/JuliaLang/julia/pull/58315
3. In the team settings for the `@JuliaLang/committers-available-to-review` team, in the "Code review" tab of the settings (https://github.com/orgs/JuliaLang/teams/committers-available-to-review/edit/review_assignment), enable GitHub's built-in "auto-assign" functionality. Here is a screenshot of the settings as I have currently configured them:

<img width="990" alt="Screenshot 2025-05-05 at 1 22 30 AM" src="https://github.com/user-attachments/assets/8309a360-13fa-429f-bdf6-792c44dd2502" />

Here are the two options for "routing algorithm":

<img width="315" alt="Screenshot: Routing algorithms" src="https://github.com/user-attachments/assets/179fdcb4-d417-47a3-9a55-b8889f0db805" />

Currently, I've selected "round robin", but we can easily switch that.

With this setup, once XXX has been merged, whenever a PR to `JuliaLang/julia` is opened, GitHub will use our specified routing algorithim to assign the PR to one member of the `@JuliaLang/committers-available-to-review` team.

I think this approach would be good, because we don't need to maintain our own GitHub Actions script, and we don't need to e.g. deal with permissions issues for `GITHUB_TOKEN`.

---

Oh, also, with this approach, we can connect the `@JuliaLang/committers-available-to-review` team to a Slack channel (https://github.com/orgs/JuliaLang/teams/committers-available-to-review/settings/reminders) and set up automatic reminders about PRs waiting for review. I don't think we need to necessarily set that up right now, but it's nice to have the option.

DilumAluthge commented:
One other quick note for now: with GitHub's built-in "auto-assign" functionality, there are two different ways that a user can take themselves out of the pool:
1. Remove themselves from the `@JuliaLang/committers-available-to-review` team.
2. Mark their GitHub "status" as "Busy" (click on your profile picture in the top-right-hand corner of any GitHub.com page, and then click on the "Set status" option right below your GitHub username.

So users can use option 2 to easily temporarily take themselves out of the pool, and then easily add themselves back (by unsetting their "Busy" GitHub status when they are ready to participate in reviews again).

LilithHafner commented:
Implemented by #58315

xlxs4 commented:
Thank you all for working on this!

DilumAluthge commented on .github/workflows/AssignCommitter.yml:
```diff
@@ -0,0 +1,47 @@
+jobs:
+  assign-reviewer:
+    runs-on: ubuntu-latest
+    if: ${{ github.event.pull_request.draft != true }}
```
We don't want this action to run in people's forks of Julia:

```suggestion
    if: ${{ github.repository == 'JuliaLang/julia' && github.event.pull_request.draft != true }}
```

DilumAluthge commented on .github/workflows/AssignCommitter.yml:
```diff
@@ -0,0 +1,47 @@
+        with:
+          script: |
+            if (context.payload.pull_request.assignees.length === 0) {
+              // Find suitable assignees
```
For this section ("Find suitable assignees"), can you implement the following?

1. Use the GitHub API to retrieve the contents of this file: https://github.com/JuliaLang/pr-assignment/blob/main/users.txt
2. Once you have the file contents, I'd recommend processing it like this.
    1. Split into lines.
    2. Strip trailing whitespace.
    3. Ignore all lines that start with `#`
    4. Strip comments, e.g. if a line is `@myname # some comment`, discard everything after `#`
    5. Require that usernames have the `@` at the beginning, for consistency with how GitHub `CODEOWNERS` files behave.

DilumAluthge replied:
Can you also add a comment in this YAML file that says something along the lines of this:

> Note to JuliaLang committers: If you would like to add yourself to the pool of committers that are available to be assigned to triage new PRs, please make a PR to this repo: https://github.com/JuliaLang/pr-assignment

DilumAluthge replied:
Something like this regex should work for parsing the username lines in `users.txt`: `^@([A-Za-z0-9\-]+?)$`

Taken from https://github.com/JuliaLang/pr-assignment/blob/b9a8017d33574462b9a502b279907d9b8e003554/.github/workflows/lint.yml#L51-L51

LilithHafner replied:
Done

DilumAluthge commented on .github/workflows/AssignCommitter.yml:
```diff
@@ -0,0 +1,47 @@
+  pull_request:
+    types: [opened, reopened, ready_for_review]
+
+permissions:
+  pull-requests: write
```
```suggestion
# Permissions for the `GITHUB_TOKEN`:
permissions:
  pull-requests: write # Needed in order to assign a user as the PR assignee
```

DilumAluthge commented on .github/workflows/AssignCommitter.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: Add Assignee
+        uses: actions/github-script@v7
+        with:
+          script: |
```
```suggestion
          retries: 5 # retry GitHub API requests up to 5 times, with exponential backoff
          script: |
```

LilithHafner committed:
Update .github/workflows/AssignCommitter.yml

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

DilumAluthge committed:
CI: `AssignCommitter.yml`: Use the `on: pull_request_target` event (not the `on: pull_request` event) (#58384)

LilithHafner committed:
Update .github/workflows/AssignCommitter.yml

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

LilithHafner committed:
Update .github/workflows/AssignCommitter.yml

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

LilithHafner committed:
Update workflow to use external reviwer list and repo collaborator check

LilithHafner committed:
Merge branch 'master' into LilithHafner-patch-1

LilithHafner commented:
@DilumAluthge, `actions/github-script@v7` is whitelisted but `actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea` is not. See https://github.com/JuliaLang/julia/actions/runs/15031350693 for log. Would you please adjust this?

DilumAluthge commented:
> @DilumAluthge, `actions/github-script@v7` is whitelisted but `actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea` is not. See https://github.com/JuliaLang/julia/actions/runs/15031350693 for log. Would you please adjust this?

Yep I can do so.

DilumAluthge commented:
> > @DilumAluthge, `actions/github-script@v7` is whitelisted but `actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea` is not. See https://github.com/JuliaLang/julia/actions/runs/15031350693 for log. Would you please adjust this?
> 
> Yep I can do so.

Done.

DilumAluthge commented on .github/workflows/AssignCommitter.yml:
```diff
@@ -0,0 +1,97 @@
+                issue_number: context.payload.pull_request.number,
+                labels: ['status: waiting for PR reviewer'],
+              });
+              }
```
I think we have one too many `}` here.

```suggestion
```

DilumAluthge committed:
Try to fix syntax

DilumAluthge committed:
Base64-decode the contents

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Merge branch 'master' into LilithHafner-patch-1

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update AssignCommitter.yml

DilumAluthge committed:
Update and rename AssignCommitter.yml to PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Fix whitespace

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Remove `\n` at end of regex

We `.trim()` each line in the previous step, so we don't need a newline character at the end of our regex.

DilumAluthge committed:
New regex

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Fix whitespace

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Update PrAssignee.yml

DilumAluthge committed:
Merge branch 'master' into LilithHafner-patch-1

DilumAluthge commented on .github/workflows/AssignCommitter.yml:
```diff
@@ -0,0 +1,47 @@
+              });
+              const member_logins = members.map(member => member.login);
+
+              // Skip PRs authored by committers
```
This has been implemented now.

DilumAluthge committed:
More fixes

LilithHafner committed:
Remove TODO (we can keep the debugging feature in "productio") & Fix regex bug

LilithHafner commented:
The logging is a bit excessive to the point where it makes the script itself harder to follow, but that's not actually a big problem.

It looks good in tests, I think it's fine to merge and see how it goes in practice. Worst case it turns out like #58315 which wasn't too bad.

DilumAluthge commented:
I'll do one last review pass later today, and then we can merge it.

DilumAluthge committed:
Merge branch 'master' into LilithHafner-patch-1

DilumAluthge reviewed (approved):
I think this is ready to merge.

DilumAluthge commented:
@LilithHafner I did one last review pass. I think this is ready to merge.

LilithHafner committed:
Update .github/CODEOWNERS

Co-authored-by: Andy Dienes <51664769+adienes@users.noreply.github.com>

LilithHafner commented on .github/CODEOWNERS:
```diff
@@ -4,6 +4,7 @@ CODEOWNERS @JuliaLang/github-actions
 
 /.github/workflows/rerun_failed.yml @DilumAluthge
 /.github/workflows/statuses.yml @DilumAluthge
+/.github/workflows/AssignCommitter.yml @LilithHafner @DilumAluthge
```
Thanks!
